server/src/main.rs

```diff
@@ -1,56 +1,90 @@
-        match error.code() {
-            git2::ErrorCode::NotFound => AppError::ObjectNotFound(error.to_string()),
-            _ => AppError::GitError(error),
-        }
-    }
-}
+const DEFAULT_REPO_ROOT: &str = "./";
+const DEFAULT_REFERENCE: &str = "heads/master";
 
 fn main() {
     let args = parse_args().get_matches();
 
     let host = args.value_of("host").unwrap_or(DEFAULT_HOST);
     let port = args.value_of("port").unwrap_or(DEFAULT_PORT);
+    let repo_root = Path::new(args.value_of("repo-root").unwrap_or(DEFAULT_REPO_ROOT));
 
-    create_server(host, port).run();
+    run_server(host, port, repo_root);
 }
 
-fn create_server(host: &str, port: &str) -> server::HttpServer<App<()>, fn() -> App<()>> {
+#[derive(Deserialize)]
+pub struct PathParams {
+    pub repo: String,
+}
 
-    let mut listenfd = ListenFd::from_env();
-    let server: server::HttpServer<App<()>, fn() -> App<()>> = server::new(|| {
-        App::new()
-            .resource("/repo/{repo}", |r| r.h(RepoHandler::new(DEFAULT_REPO_PATH.to_string())))
-    });
+#[derive(Deserialize)]
+pub struct QueryParams {
+    pub reference: Option<String>,
+    pub file: String,
+}
 
-    match listenfd.take_tcp_listener(0).expect("can't take tcp listener") {
-        Some(l) => server.listen(l),
-        None => {
-            let address = format!("{}:{}", host, port);
-            println!("Listening to {}", address);
-            server.bind(address).expect("can't bind into address")
-        }
-    }
+pub struct AppState {
+    pub git_repos: Addr<GitRepos>,
+}
+
+fn run_server(host: &str, port: &str, repo_root: &Path) {
+    let _sys = System::new("gitkv-server");
+
+    let addr = GitRepos::new(git::load_repos(&repo_root)).start();
+
+    let listen_address = format!("{}:{}", host, port);
+
+    server::new(move || {
+        App::with_state(AppState {
+            git_repos: addr.clone(),
+        })
+        .middleware(middleware::Logger::default())
+        .resource("/repo/{repo}", |r| r.method(http::Method::GET).f(get_repo))
+    })
+    .bind(listen_address)
+    .expect("can't bind into address")
+    .run();
+}
+
+fn get_repo(req: &HttpRequest<AppState>) -> impl Responder {
+    //TODO https://actix.rs/docs/errors/
+    let path_params = actix_web::Path::<PathParams>::extract(req).expect("Wrong path params");
+    let query_params = actix_web::Query::<QueryParams>::extract(req).expect("Wrong query params");
+    let repo_key = path_params.repo.to_string();
+    let filename = query_params.file.to_string();
+    let reference = format!(
+        "refs/{}",
+        query_params
+            .reference
+            .as_ref()
+            .unwrap_or(&DEFAULT_REFERENCE.to_string())
+    );
+    let gr: &Addr<GitRepos> = &req.state().git_repos;
+    //TODO return proper content type depending on the content of the blob
+    gr.send(CatFile {
+        repo_key,
+        filename,
+        reference,
+    })
+    .map(|x| {
+        x.0.map(Binary::from)
```

Comment:
Don't mind if we fix in another PR — but would be inclined to use a struct here instead of a tuple, `x.0` is kind of gross because it's hard to see from the code what it is that is being operated on.

Reply:
I'm usually anti-tuple so happy to refactor now that you've called me out on it 😛

Reply:
Wait, got mixed up, was thinking of the tuple I used here. Though in this case it's necessary for building the map.
https://github.com/intenthq/gitkv/blob/de7bf870644607c2bfd7010ed0b350ed8db7e5bd/git/src/lib.rs#L48

Given the sprint ends tomorrow I'll pick this up in a future PR?